### PR TITLE
UI and flow updates

### PR DESCRIPTION
**Pull request by mrdarrengriffin**
- Updated copy
- Improved the overall UI by adding a logo, demo graphic and Casita
- Improved the card UX by adding card actions and chevrons
- Added instruction page with clearer steps for users to follow
- Added catch for microphone permission request
- Added copy to clipboard button

### CODE REVIEW

File: src/components/card-layout.ts [hunk 1]
```diff
@@ -1,19 +1,39 @@
...
   @property() public header!: string;
+  @property() public casita!: string;
+  @property() public alignment: "center" | "left" = "left";
 
   render() {
+    let cardHeader = null;
+    if (this.header) {
+      cardHeader = html`
+        <h1 class="card-header">
+          ${this.casita
```
Comment (balloob) [inline src/components/card-layout.ts]: This is better done with a slot. `card-layout` is a generic component and shouldn't have the notion of casitas.

File: src/components/card-layout.ts [hunk 2]
```diff
@@ -1,19 +1,39 @@
...
 @customElement("card-layout")
 export class CardLayout extends LitElement {
   @property() public header!: string;
+  @property() public casita!: string;
+  @property() public alignment: "center" | "left" = "left";
 
   render() {
+    let cardHeader = null;
+    if (this.header) {
+      cardHeader = html`
```
Comment (balloob) [inline src/components/card-layout.ts]: I would just inline this in the html below. It will be small with casita replaced by a named slot

File: src/components/error-notice.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      </div>
+    `;
+  }
+
+  static styles = css`
+    :host {
+      display: flex;
+      flex-direction: column;
+      align-items: center;
+      max-width: 450px;
```
Comment (balloob) [inline src/components/error-notice.ts]: it's a bit weird there is a max width on a `:host` selector. It limits reusability. That should be added to the parent.

File: src/components/error-notice.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      width: 100%;
```
Comment (balloob) [inline src/components/error-notice.ts]: I don't think this is needed, flex elements already take the whole width?

File: src/components/error-notice.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    .notice {
+      width: 100%;
```
Comment (balloob) [inline src/components/error-notice.ts]: div is a block element, already full width.

File: src/components/share-url.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+  render() {
+    return html`
+      <md-text-button
+        trailing-icon
+        @click=${this.copyToClipboard}
+        ?disabled=${this.copied}
+      >
+        ${this.copied ? "URL copied to clipboard" : "Copy URL to clipboard"}
+        ${this.copied ? ICON_CHECK_SLOTTED : ICON_COPY_SLOTTED}
```
Comment (balloob) [inline src/components/share-url.ts]: combine these
```suggestion
        ${this.copied ?
           html`URL copied to clipboard ${ICON_CHECK_SLOTTED}" :
           html`Copy URL to clipboard ${ICON_COPY_SLOTTED}`
        }
```

File: src/components/share-url.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  async copyToClipboard() {
+    if (this.copied) return;
+
+    await navigator.clipboard.writeText(
+      "https://ohf-voice.github.io/wake-word-collective/",
+    );
+
+    this.copied = true;
+    this.requestUpdate("copied");
```
Comment (balloob) [inline src/components/share-url.ts]: If you need to call this, it means you forgot to add a `@state` decorator on the property :)

File: src/components/share-url.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    setTimeout(() => {
+      this.copied = false;
+      this.requestUpdate("copied");
```
Comment (balloob) [inline src/components/share-url.ts]: Same

File: src/components/share-url.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }, 5000);
+  }
+
+  static styles = css`
+    md-text-button {
+      margin-top: 16px;
```
Comment (balloob) [inline src/components/share-url.ts]: Margins should never be part of a component itself. It should always be applied by the component that includes them. This improves reusability.

File: src/components/share-url.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      width: 100%;
```
Comment (balloob) [inline src/components/share-url.ts]: What type of element is md-text-button? Shouldn't we just do display: block?

File: src/pages/consent/consent-page.ts [hunk 11]
```diff
@@ -30,28 +25,32 @@ export class ConsentPage extends LitElement {
     this.descriptionField.value = this.description;
   }
 
+  getChangeLink() {
```
Comment (balloob) [inline src/pages/consent/consent-page.ts]: let's inline this. Extracting this makes code harder to read.

File: src/pages/consent/consent-page.ts [hunk 12]
```diff
@@ -30,28 +25,32 @@ export class ConsentPage extends LitElement {
...
+  }
+
   render() {
     if (this.givingConsent) {
       return html`
-        <card-layout header="Starting recording studio">
+        <card-layout header="Checking audio device">
           <div class="card-content">
-            <i>Please wait.</i>
+            <i>Please wait...</i>
```
Comment (balloob) [inline src/pages/consent/consent-page.ts]:
```suggestion
            <i>Please wait…</i>
```

File: src/pages/consent/consent-page.ts [hunk 13]
```diff
@@ -30,28 +25,32 @@ export class ConsentPage extends LitElement {
     this.descriptionField.value = this.description;
   }
 
+  getChangeLink() {
+    if (Object.keys(WAKE_WORDS).length < 2) return;
+    return html`&nbsp;(<a href="#" @click=${this.cancelConsent}>change</a>)`;
+  }
+
   render() {
     if (this.givingConsent) {
```
Comment (balloob) [inline src/pages/consent/consent-page.ts]: We can remove this boolean, as this is now on the next page.

File: src/pages/instructions/instructions-page.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+  @property() public startRecording!: (recorder: Recorder) => void;
+
+  @property() public description!: string;
+
+  @property() public wakeWord!: string;
+
+  @property({
+    attribute: false,
+  })
+  public recorderError: boolean = false;
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]: You want this to be `@state` I think? It's not set externally. 
```suggestion
  @state
  private recorderError = false;
```

File: src/pages/instructions/instructions-page.ts [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+import "@material/web/list/list-item";
+import "@material/web/icon/icon";
+import "../../components/card-layout";
+import "../../components/error-notice";
+import { createRecorder } from "../../util/recorder";
+import type { Recorder } from "../../util/recorder";
+
+@customElement("instructions-page")
+export class InstructionsPage extends LitElement {
+  @property() public startRecording!: (recorder: Recorder) => void;
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]: this one should be `attribute: false`

File: src/pages/instructions/instructions-page.ts [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        <img src="./images/demo.gif" class="demo" />
+        <div class="card-content">
+          ${this.recorderError
+            ? html`
+                <error-notice>
+                  There was an issue accessing your microphone. Please check
+                  your browser permissions and try again
+                </error-notice>
+              `
+            : null}
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]:
```suggestion
            : nothing}
```

File: src/pages/landing/landing-page.ts [hunk 17]
```diff
@@ -2,86 +2,89 @@ import { LitElement, css, html } from "lit";
...
-          ${Object.entries(WAKE_WORDS).map(
-            ([key, label]) => html`
-              <md-list-item
-                type="button"
-                @click=${() => {
-                  this.selectWakeWord(key);
-                }}
+        ${Object.entries(WAKE_WORDS).map(
+          ([key, label]) => html`
+            <div class="card-actions">
```
Comment (balloob) [inline src/pages/landing/landing-page.ts]: Don't put card-actions inside the list. That element is not meant to be rendered multiple times

File: src/pages/instructions/instructions-page.ts [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        line-height: 1.5;
+      }
+
+      .note p:first-of-type {
+        margin-top: 0;
+      }
+
+      ol li:not(:last-child) {
+        margin-bottom: 0.5rem;
+      }
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]: these things are duplicate with `introduction-page`, no? Should we move them to `PAGE_STYLES` ?

File: static/index.html [hunk 19]
```diff
@@ -1,140 +1,134 @@
...
+  <title>Wake Word Collective - Open Home Foundation</title>
+  <link rel="preconnect" href="https://fonts.googleapis.com">
+  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
+  <link
+    href="https://fonts.googleapis.com/css2?family=Figtree:ital,wght@0,300..900;1,300..900&family=Instrument+Sans:ital,wght@0,400..700;1,400..700&display=swap"
+    rel="stylesheet">
+  <link rel="icon" href="/images/favicon.png" type="image/png" />
+  <link rel="preload" as="image" href="./images/logo.svg">
+  <link rel="preload" as="image" href="./images/demo.gif">
+  <link rel="preload" as="image" href="./images/casita/heart.gif">
```
Comment (balloob) [inline static/index.html]: preload should be used right away. These are not used for a long time.

File: static/index.html [hunk 20]
```diff
@@ -1,140 +1,134 @@
...
+      --ha-card-border-radius: 12px;
+    }
 
-        --md-sys-color-primary: #03a9f4;
-        --md-sys-color-on-primary: #fff;
-        --md-ref-typeface-brand: var(--roboto-font);
-        --md-ref-typeface-plain: var(--roboto-font);
+    * strong {
+      color: red !important;
+    }
```
Comment (balloob) [inline static/index.html]: This was testing :) 
```suggestion
```

Comment (mrdarrengriffin) [reply static/index.html]: 🤦🏼

File: src/components/card-layout.ts [hunk 21]
```diff
@@ -1,19 +1,28 @@
-import { LitElement, css, html } from "lit";
+import { LitElement, css, html, nothing } from "lit";
 import { customElement, property } from "lit/decorators.js";
+import { PAGE_STYLES } from "../const";
 
 @customElement("card-layout")
 export class CardLayout extends LitElement {
   @property() public header!: string;
+  @property() public casita!: string;
```
Comment (balloob) [inline src/components/card-layout.ts]: No longer used
```suggestion
```

File: src/components/header-logo.ts [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+      <header>
+        <img src="./images/logo.svg" alt="Wake Word Collective logo" />
+      </header>
+    `;
+  }
+
+  static styles = css`
+    :host {
+      flex: 0 !important;
+    }
```
Comment (balloob) [inline src/components/header-logo.ts]: Why? If this is to reset some style from the parent, then the parent should reset it.

File: src/components/header-logo.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+      margin: 0 auto;
+      max-width: 450px;
+      text-align: center;
+      margin-top: 50px;
+      padding: 0 16px;
+    }
+
+    img {
+      width: 100%;
+      height: auto;
```
Comment (balloob) [inline src/components/header-logo.ts]: display: block ?

File: src/components/share-url.ts [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+@customElement("share-url")
+export class ShareURL extends LitElement {
+  @state() private copied: boolean = false;
+
+  render() {
+    return html`
+      <md-text-button
+        trailing-icon
+        @click=${this.copyToClipboard}
+        ?disabled=${this.copied}
```
Comment (balloob) [inline src/components/share-url.ts]: It's also a property, so you can also set it like this:
```suggestion
        .disabled=${this.copied}
```

Comment (balloob) [reply src/components/share-url.ts]: (properties are faster, as you bypass the DOM model and directly talk to the class)

File: src/components/share-url.ts [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+      >
+        ${this.copied
+          ? html`URL copied to clipboard! ${ICON_CHECK_SLOTTED}`
+          : html`Copy URL to clipboard ${ICON_COPY_SLOTTED}`}
+      </md-text-button>
+    `;
+  }
+
+  async copyToClipboard() {
+    if (this.copied) return;
```
Comment (balloob) [inline src/components/share-url.ts]: I don't think this is necessary. Buttons that are disabled prevent all click events from happening.

File: src/const.ts [hunk 26]
```diff
@@ -1,3 +1,45 @@
...
+export const PAGE_STYLES = css`
+  strong {
+    font-weight: 600;
+  }
+
+  ol li:not(:last-child) {
+    margin-bottom: 0.5rem;
+  }
+
+  .casita img {
```
Comment (balloob) [inline src/const.ts]: Is a casita image common enough ?

File: src/pages/consent/consent-page.ts [hunk 27]
```diff
@@ -31,27 +27,32 @@ export class ConsentPage extends LitElement {
   }
 
   render() {
+    preloadImage("./images/demo.gif");
```
Comment (balloob) [inline src/pages/consent/consent-page.ts:25]: Do not put this in `render`. Put this in `firstUpdated`

Comment (balloob) [reply src/pages/consent/consent-page.ts:25]: `render` is called on every render. FirstUpdated is only called once, but only after the component has rendered, so the user can interact with it.

File: src/components/share-url.ts (hunk 25 above)
Comment (mrdarrengriffin) [reply src/components/share-url.ts]: I put this here as the function was async so there might have been a chance of it calling twice if someone clicked fast enough. However, I agree, overkill solution

File: src/pages/consent/consent-page.ts (hunk 13 above)
Comment (balloob) [reply src/pages/consent/consent-page.ts]: You marked this one as resolved, but `givingConsent` is still not used? Shouldn't this whole block be removed. It was only there to ensure that the page changed while we were requesting microphone access

File: src/pages/instructions/instructions-page.ts [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+      return;
+    }
+
+    this.startRecording(recorder);
+  }
+
+  static styles = [
+    PAGE_STYLES,
+    css`
+      a {
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]: Surprised this one isn't in `PAGE_STYLES` :D

File: src/pages/instructions/instructions-page.ts [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+        border-radius: 12px;
+      }
+
+      svg {
+        width: 24px;
+      }
+
+      img.demo {
+        width: 100%;
+      }
```
Comment (balloob) [inline src/pages/instructions/instructions-page.ts]: Don't you want to just show it as a block element?
```suggestion
      img.demo {
        display: block;
      }
```

Comment (mrdarrengriffin) [reply src/pages/instructions/instructions-page.ts]: I'll leave width in as it still overflows the card (the parent is flex)

File: src/pages/recording/recording-page.ts [hunk 30]
```diff
@@ -37,6 +38,8 @@ export class RecordingPage extends LitElement {
   }
 
   render() {
+    preloadImage("./images/casita/heart.gif");
```
Comment (balloob) [inline src/pages/recording/recording-page.ts]: put this in firstUpdated

File: src/pages/recording/recording-page.ts [hunk 31]
```diff
@@ -13,9 +14,11 @@ export class RecordingPage extends LitElement {
     attribute: false,
   })
   public recorder!: Recorder;
-  @state() public times = 10;
+  @state() private times = 10;
 
   protected firstUpdated(changedProperties: PropertyValues): void {
+    preloadImage("./images/casita/heart.gif");
```
Comment (balloob) [inline src/pages/recording/recording-page.ts:22]: Always call super first :)

File: src/pages/recording/recording-page.ts [hunk 32]
```diff
@@ -74,61 +72,65 @@ export class RecordingPage extends LitElement {
...
+  static styles = [
+    PAGE_STYLES,
+    css`
+      :host {
+        display: flex;
+        flex-direction: column;
+        align-items: center;
+        width: 100%;
+        height: 100%;
+        flex: 1;
```
Comment (balloob) [inline src/pages/recording/recording-page.ts]: Is this to control how this component is rendered in the host? If so, it should move to the host.

File: src/pages/thank-you-page/thank-you-page.ts [hunk 33]
```diff
@@ -1,32 +1,60 @@
...
+          <md-text-button has-icon trailing-icon @click=${this.restart}>
+            Record another sample ${ICON_CHEVRON_SLOTTED}
+          </md-text-button>
         </div>
       </card-layout>
     `;
   }
+
+  restart() {
+    return (window.location.href = window.location.href.replace(
```
Comment (balloob) [inline src/pages/thank-you-page/thank-you-page.ts]: why return it?

Comment (mrdarrengriffin) [reply src/pages/thank-you-page/thank-you-page.ts]: Good question

File: src/pages/consent/consent-page.ts [hunk 34]
```diff
@@ -96,47 +86,40 @@ export class ConsentPage extends LitElement {
...
-      padding-left: 16px;
-    }
-  `;
+  static styles = [
+    PAGE_STYLES,
+    css`
+      a {
+        color: var(--md-sys-color-primary);
+      }
+
```
Comment (balloob) [inline src/pages/consent/consent-page.ts]: Not needed
```suggestion
```

File: src/pages/landing/landing-page.ts [hunk 35]
```diff
@@ -2,86 +2,89 @@ import { LitElement, css, html } from "lit";
...
-    a {
-      color: var(--md-sys-color-primary);
-    }
+  static styles = [
+    PAGE_STYLES,
+    css`
+      a {
+        color: var(--md-sys-color-primary);
+      }
+
```
Comment (balloob) [inline src/pages/landing/landing-page.ts]: Not needed
```suggestion
```

File: src/pages/landing/landing-page.ts [hunk 36]
```diff
@@ -2,86 +2,89 @@ import { LitElement, css, html } from "lit";
...
+      svg {
+        width: 24px;
+      }
+
+      p:last-child {
+        margin-bottom: 0;
+      }
+
```
Comment (balloob) [inline src/pages/landing/landing-page.ts]:
```suggestion
```

File: src/pages/page-router.ts [hunk 37]
```diff
@@ -82,8 +110,10 @@ export class PageRouter extends LitElement {
   }
 
   static styles = css`
-    * {
-      flex: 1;
+    recording-page {
+      width: 100%;
+      height: 100%;
+      align-items: center;
```
Comment (balloob) [inline src/pages/page-router.ts]: Are you trying to align the items inside the `recording-page`? That seems like something that should be defined inside `recording-page` ?

**Review by balloob — APPROVED** [review]

File: src/pages/page-router.ts [hunk 38]
```diff
@@ -38,6 +42,7 @@ export class PageRouter extends LitElement {
   render() {
     if (!this.wakeWord) {
       return html`
+        <header-logo></header-logo>
```
Comment (balloob) [inline src/pages/page-router.ts]: Move this into `landing-page`

File: src/pages/page-router.ts [hunk 39]
```diff
@@ -48,25 +53,48 @@ export class PageRouter extends LitElement {
     }
 
     if (this.wakeWord === "thank_you") {
-      return html`<thank-you-page></thank-you-page>`;
+      return html`
+        <header-logo></header-logo>
```
Comment (balloob) [inline src/pages/page-router.ts]: move this into thank-you-page

File: src/pages/page-router.ts [hunk 40]
```diff
@@ -48,25 +53,48 @@ export class PageRouter extends LitElement {
...
+        <thank-you-page></thank-you-page>
+      `;
     }
 
-    if (!this.recorder) {
+    if (this.showInstructions) {
       return html`
-        <consent-page
+        <header-logo></header-logo>
```
Comment (balloob) [inline src/pages/page-router.ts]: move into instructions-page

File: src/pages/page-router.ts [hunk 41]
```diff
@@ -48,25 +53,48 @@ export class PageRouter extends LitElement {
...
+            this.showInstructions = false;
+            this.recorder?.start();
+          }}
+        ></instructions-page>
+      `;
+    }
+
+    if (!this.recorder) {
+      return html`
+        <header-logo></header-logo>
```
Comment (balloob) [inline src/pages/page-router.ts]: Move into consent-page

File: src/pages/page-router.ts [hunk 42]
```diff
@@ -82,8 +104,9 @@ export class PageRouter extends LitElement {
   }
 
   static styles = css`
-    * {
-      flex: 1;
+    recording-page {
+      width: 100%;
+      height: 100%;
```
Comment (balloob) [inline src/pages/page-router.ts]: You can revert this change again.

File: src/pages/page-router.ts [hunk 43]
```diff
@@ -51,22 +55,40 @@ export class PageRouter extends LitElement {
...
             });
+
+            this.showInstructions = false;
+            this.recorder?.start();
+          }}
+        ></instructions-page>
+      `;
+    }
+
+    if (!this.recorder) {
```
Comment (balloob) [inline src/pages/page-router.ts]: So this check is now a bit weird. The recorder is created inside instructions page. 

We might have to switch the router to just a state parameter now that we have multiple stages

Comment (mrdarrengriffin) [reply src/pages/page-router.ts]: Hmm, I'm not sure how to best implement that
